File: 백준/DFS, BFS/2573.js
const fs = require("fs");

const inputs = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const [n, m] = inputs[0].split(" ").map((e) => Number(e));
const graph = inputs.slice(1).map((e) => e.split(" ").map((e) => Number(e)));
const dx = [1, 0, -1, 0];
const dy = [0, 1, 0, -1];

class Queue {
  constructor() {
    this.q = [];
    this.head = 0;
    this.rear = 0;
  }

  isEmpty() {
    return this.head === this.rear;
  }

  push(value) {
    this.q.push(value);
    this.rear++;
  }

  shift() {
    let rv = this.q[this.head];
    delete this.q[this.head++];
    return rv;
  }
}

const bfs = (start, checked) => {
  const q = new Queue();
  q.push(start);

  checked[start[0]][start[1]] = 1;

  while (!q.isEmpty()) {
    const [cy, cx] = q.shift();

    dx.forEach((e, i) => {
      let nx = cx + e;
      let ny = cy + dy[i];

      if (
        nx >= 0 &&
        nx < m &&
        ny >= 0 &&
        ny < n &&
        !checked[ny][nx] &&
        graph[ny][nx]
      ) {
        checked[ny][nx] = 1;
        q.push([ny, nx]);
      }
    });
  }
};

const melt = (graph) => {
  const dh = Array.from({ length: n }, () =>
    Array.from({ length: m }, () => 0)
  );
  let isMelted = true;

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (graph[i][j]) {
        dx.forEach((e, index) => {
          let nx = j + e;
          let ny = i + dy[index];

          if (nx >= 0 && nx < m && ny >= 0 && ny < n && !graph[ny][nx]) {
            dh[i][j]++;
          }
        });
      }
    }
  }

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      graph[i][j] -= dh[i][j];
      graph[i][j] = Math.max(0, graph[i][j]);

      if (graph[i][j]) {
        isMelted = false;
      }
    }
  }

  return isMelted;
};

let flag = false;
let answer = 0;
while (!melt(graph)) {
  let cnt = 0;
  const checked = Array.from({ length: n }, () =>
    Array.from({ length: m }, () => 0)
  );
  answer++;

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (graph[i][j] && !checked[i][j]) {
        bfs([i, j], checked);
        cnt++;
      }
    }
  }

  if (cnt > 1) {
    flag = true;
    break;
  }
}

if (flag) {
  console.log(answer);
} else {
  console.log(0);
}

File: 백준/DFS, BFS/2573.py
import sys
from collections import deque

n, m = map(int, sys.stdin.readline().split())
graph = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]

dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]

def count(start):
    cnt = 0
    [cy, cx] = start
    for i in range(len(dx)):
        ny = cy + dy[i]
        nx = cx + dx[i]
        
        if graph[ny][nx] == 0:
            cnt += 1
    return cnt

def bfs(start):
    [sy, sx] = start
    q = deque()
    visited[sy][sx] = True

    q.append(start)
    
    while q:
        [cy, cx] = q.popleft()
        to_melt[cy][cx] = count([cy, cx])
        
        for i in range(len(dx)):
            ny = cy + dy[i]
            nx = cx + dx[i]

            if nx >= 0 and nx < m and ny >= 0 and ny < n and not visited[ny][nx] and graph[ny][nx] != 0:
                q.append([ny, nx])
                visited[ny][nx] = True

    return 1
  

ans = 0
flag = False
while True:
    visited = [[False] * m for _ in range(n)]
    to_melt = [[0 for _ in range(m)] for _ in range(n)]
    cnt = 0
    
    for i in range(n):
        for j in range(m): 
            if graph[i][j] != 0 and not visited[i][j]:
                cnt += bfs([i, j])

                if cnt > 1:
                    break
        else:
            continue
        break
    
    else:
        if cnt == 0:
            flag = True
            break
        for i in range(n):
            for j in range(m):
                graph[i][j] = max(0, graph[i][j] - to_melt[i][j])
        ans += 1
        continue
    break

if flag:
    print(0)
else:
    print(ans)